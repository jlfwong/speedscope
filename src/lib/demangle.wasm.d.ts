interface WasmDemangleModule {
  wasm_demangle(mangled: string): string
}

export default function ModuleFactory(options?: unknown): Promise<WasmDemangleModule>
