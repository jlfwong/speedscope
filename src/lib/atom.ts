// This is a very small application state management library.
//
// All it provides is a way of specifying state "atoms" (basically a value with
// a setter and a way to be notified when the value updates), and a single
// preact hook to manage the subscribe/unsubscribe process for you.
//
// At the moment, atoms are intended to be globally defined, but the system
// could easily be adapted to pass down the atoms via context rather than being
// globally available.
//
// To support complex data-types, you can either have simple functions that call
// the public setter, or subclass Atom. So, for example, here are a few
// different ways of making a list atom with convenient helpers:
//
// # Simple functions being passed atom
//
//     const myListAtom = new Atom<number[]>([], "myList")
//     function push<T>(atom: Atom<T[]>, t: T) {
//       const next = [...atom.get(), t]
//       atom.set(next)
//     }
//     function removeLast<T>(atom: Atom<T[]>, t: T) {
//       const next = [...atom.get()]
//       next.pop()
//       atom.set(next)
//     }
//
//
// # Simple functions operating on global atoms
//
//     const myListAtom = new Atom<number[]>([], "myLisT")
//     function push(t: number) {
//       myListAtom.set([...myListAtom.get(), t])
//     }
//     function removeLast() {
//       const next = [...myListAtom.get()]
//       myListAtom.set(next)
//     }
//
//
// # Subclassing
//
//     class ListAtom<T> extends Atom<T[]> {
//        push(t: T) {
//          const next = [...this.state, t]
//          this.set(next)
//        }
//        removeLast() {
//          const next = [...this.state]
//          next.pop()
//          this.set(next)
//        }
//     }
//     const myListAtom = new ListAtom<number>([], "myList")
//
// This library is inspired by https://recoiljs.org/

import {useLayoutEffect, useState} from 'preact/hooks'

type AtomListener = () => void

let AtomDev: {[key: string]: Atom<any>} | null = null
let hotReloadStash: Map<string, any> | null = null

declare const module: any
if (process.env.NODE_ENV === 'development') {
  ;(window as any)['Atom'] = AtomDev = {}

  module.hot.dispose(() => {
    if (AtomDev) {
      hotReloadStash = new Map()
      for (let key in AtomDev) {
        hotReloadStash.set(key, AtomDev[key].get())
      }
    }

    ;(window as any)['Atom_hotReloadStash'] = hotReloadStash
  })

  hotReloadStash = (window as any)['Atom_hotReloadStash'] || null
}

export class Atom<T> {
  private observers: AtomListener[] = []
  constructor(protected state: T, debugKey: string) {
    if (process.env.NODE_ENV === 'development') {
      if (hotReloadStash?.has(debugKey)) {
        // If we have a stored value from a previous hot reload, use that
        // instead of whatever was passed to the constructor.
        this.state = hotReloadStash.get(debugKey)
      }

      if (AtomDev) {
        if (debugKey in AtomDev) {
          console.warn(`[Atom] Multiple atoms tried to register with the key ${debugKey}`)
        }
        AtomDev[debugKey] = this
      }
    }

    // We do the bind here rather than in the definition to facilitate
    // inheritance (we want the value defined on both the prototype and the
    // instance).
    this.set = this.set.bind(this)
    this.get = this.get.bind(this)
  }

  set(t: T) {
    if (this.state === t) {
      // No-op if the value didn't actually change
      return
    }
    this.state = t
    this.observers.forEach(fn => fn())
  }

  get(): T {
    return this.state
  }

  subscribe(listener: AtomListener) {
    this.observers.push(listener)
  }

  unsubscribe(listener: AtomListener) {
    const index = this.observers.indexOf(listener)
    if (index !== -1) {
      this.observers.splice(index, 1)
    }
  }
}

export function useAtom<T>(atom: Atom<T>): T {
  const [value, setValue] = useState(atom.get())

  useLayoutEffect(() => {
    // We need to setValue here because it's possible something has changed the
    // value in the store between the atom.get() call above and layout. In most
    // cases this should no-op.
    setValue(atom.get())

    function listener() {
      setValue(atom.get())
    }

    atom.subscribe(listener)
    return () => {
      atom.unsubscribe(listener)
    }
  }, [atom])

  return value
}
