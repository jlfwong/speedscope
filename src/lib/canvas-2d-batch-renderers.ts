// This file contains a collection of classes which make it easier to perform
// batch rendering of Canvas2D primitives. The advantage of this over just doing
// ctx.beginPath() ... ctx.rect(...) ... ctx.endPath() is that you can construct
// several different batch renderers are the same time, then decide on their
// paint order at the end.
//
// See FlamechartPanZoomView.renderOverlays for an example of how this is used.

export interface TextArgs {
  text: string
  x: number
  y: number
}

export class BatchCanvasTextRenderer {
  private argsBatch: TextArgs[] = []

  text(args: TextArgs) {
    this.argsBatch.push(args)
  }

  fill(ctx: CanvasRenderingContext2D, color: string) {
    if (this.argsBatch.length === 0) return
    ctx.fillStyle = color
    for (let args of this.argsBatch) {
      ctx.fillText(args.text, args.x, args.y)
    }
    this.argsBatch = []
  }
}

export interface RectArgs {
  x: number
  y: number
  w: number
  h: number
}

export class BatchCanvasRectRenderer {
  private argsBatch: RectArgs[] = []

  rect(args: RectArgs) {
    this.argsBatch.push(args)
  }

  private drawPath(ctx: CanvasRenderingContext2D) {
    ctx.beginPath()
    for (let args of this.argsBatch) {
      ctx.rect(args.x, args.y, args.w, args.h)
    }
    ctx.closePath()
    this.argsBatch = []
  }

  fill(ctx: CanvasRenderingContext2D, color: string) {
    if (this.argsBatch.length === 0) return
    ctx.fillStyle = color
    this.drawPath(ctx)
    ctx.fill()
  }

  stroke(ctx: CanvasRenderingContext2D, color: string, lineWidth: number) {
    if (this.argsBatch.length === 0) return
    ctx.strokeStyle = color
    ctx.lineWidth = lineWidth
    this.drawPath(ctx)
    ctx.stroke()
  }
}
