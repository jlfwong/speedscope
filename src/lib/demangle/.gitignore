gcc
demangle.wasm.esnext.js
