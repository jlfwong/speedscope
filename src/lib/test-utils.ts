import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
