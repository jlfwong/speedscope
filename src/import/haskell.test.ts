import {checkProfileSnapshot} from '../lib/test-utils'

test('importFromHaskell', async () => {
  await checkProfileSnapshot('./sample/profiles/haskell/simple.prof')
})
