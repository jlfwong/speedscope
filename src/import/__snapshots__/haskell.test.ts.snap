// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromHaskell 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 144,
      "line": undefined,
      "name": "MAIN.MAIN",
      "selfWeight": 0,
      "totalWeight": 798,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 56,
      "line": undefined,
      "name": "GHC.Conc.Signal.CAF",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 73,
      "line": undefined,
      "name": "GHC.IO.Encoding.CAF",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 75,
      "line": undefined,
      "name": "GHC.IO.Encoding.Iconv.CAF",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 84,
      "line": undefined,
      "name": "GHC.IO.Handle.FD.CAF",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 119,
      "line": undefined,
      "name": "Text.Printf.CAF",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 26,
      "line": undefined,
      "name": "Main.CAF:eta1_r6nI",
      "selfWeight": 0,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(21,1)-(38,42)",
      "key": 19,
      "line": undefined,
      "name": "Main.main",
      "selfWeight": 0,
      "totalWeight": 733,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(56,1)-(57,42)",
      "key": 4,
      "line": undefined,
      "name": "Main.check",
      "selfWeight": 320,
      "totalWeight": 320,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:31:9-29",
      "key": 16,
      "line": undefined,
      "name": "Main.main.long",
      "selfWeight": 0,
      "totalWeight": 3,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 28,
      "line": undefined,
      "name": "Main.CAF:eta_r6nG",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 35,
      "line": undefined,
      "name": "Main.CAF:io1",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 27,
      "line": undefined,
      "name": "Main.CAF:lvl2_r6nH",
      "selfWeight": 0,
      "totalWeight": 3,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(61,1)-(63,26)",
      "key": 3,
      "line": undefined,
      "name": "Main.make",
      "selfWeight": 406,
      "totalWeight": 409,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:63:19-26",
      "key": 1,
      "line": undefined,
      "name": "Main.make.d2",
      "selfWeight": 2,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:63:9-16",
      "key": 2,
      "line": undefined,
      "name": "Main.make.i2",
      "selfWeight": 1,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 24,
      "line": undefined,
      "name": "Main.CAF:lvl4_r6nL",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 22,
      "line": undefined,
      "name": "Main.CAF:lvl8_r6nP",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 21,
      "line": undefined,
      "name": "Main.CAF:main1",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 29,
      "line": undefined,
      "name": "Main.CAF:main11",
      "selfWeight": 0,
      "totalWeight": 4,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:27:9-35",
      "key": 15,
      "line": undefined,
      "name": "Main.main.c",
      "selfWeight": 0,
      "totalWeight": 8,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 30,
      "line": undefined,
      "name": "Main.CAF:main12",
      "selfWeight": 0,
      "totalWeight": 4,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 23,
      "line": undefined,
      "name": "Main.CAF:main5",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 31,
      "line": undefined,
      "name": "Main.CAF:main7",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 25,
      "line": undefined,
      "name": "Main.CAF:main9",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:23:9-12",
      "key": 33,
      "line": undefined,
      "name": "Main.CAF:main_maxN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:23:9-35",
      "key": 13,
      "line": undefined,
      "name": "Main.main.maxN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:22:9",
      "key": 34,
      "line": undefined,
      "name": "Main.CAF:main_n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:22:9-14",
      "key": 12,
      "line": undefined,
      "name": "Main.main.n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:24:9-16",
      "key": 32,
      "line": undefined,
      "name": "Main.CAF:main_stretchN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:24:9-27",
      "key": 14,
      "line": undefined,
      "name": "Main.main.stretchN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:17:1-4",
      "key": 36,
      "line": undefined,
      "name": "Main.CAF:minN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:17:1-8",
      "key": 9,
      "line": undefined,
      "name": "Main.minN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 146,
      "line": undefined,
      "name": "GC.GC",
      "selfWeight": 46,
      "totalWeight": 46,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 147,
      "line": undefined,
      "name": "PROFILING.OVERHEAD_of",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 145,
      "line": undefined,
      "name": "SYSTEM.SYSTEM",
      "selfWeight": 19,
      "totalWeight": 19,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:35:26-54",
      "key": 18,
      "line": undefined,
      "name": "Main.main.\\\\",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:19:1-56",
      "key": 8,
      "line": undefined,
      "name": "Main.io",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:34:9-28",
      "key": 17,
      "line": undefined,
      "name": "Main.main.vs",
      "selfWeight": 0,
      "totalWeight": 721,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(42,1)-(45,38)",
      "key": 11,
      "line": undefined,
      "name": "Main.depth",
      "selfWeight": 0,
      "totalWeight": 721,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(49,1)-(52,31)",
      "key": 7,
      "line": undefined,
      "name": "Main.sumT",
      "selfWeight": 0,
      "totalWeight": 721,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:51:9-31",
      "key": 6,
      "line": undefined,
      "name": "Main.sumT.a",
      "selfWeight": 3,
      "totalWeight": 369,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:52:9-31",
      "key": 5,
      "line": undefined,
      "name": "Main.sumT.b",
      "selfWeight": 1,
      "totalWeight": 352,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:45:9-38",
      "key": 10,
      "line": undefined,
      "name": "Main.depth.n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
  ],
  "name": "binary-trees time",
  "stacks": Array [
    "MAIN.MAIN;Main.CAF:eta1_r6nI;Main.main;Main.check 1.00ms",
    "MAIN.MAIN;Main.CAF:lvl2_r6nH;Main.main;Main.main.long;Main.make 3.00ms",
    "MAIN.MAIN;Main.CAF:main11;Main.main;Main.main.c;Main.check 4.00ms",
    "MAIN.MAIN;Main.CAF:main12;Main.main;Main.main.c;Main.make 4.00ms",
    "MAIN.MAIN;GC.GC 46.00ms",
    "MAIN.MAIN;SYSTEM.SYSTEM 19.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.check 153.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.make;Main.make.d2 2.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.make;Main.make.i2 1.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.make 210.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a 3.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b;Main.check 162.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b;Main.make 189.00ms",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b 1.00ms",
  ],
}
`;

exports[`importFromHaskell 2`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 144,
      "line": undefined,
      "name": "MAIN.MAIN",
      "selfWeight": 648,
      "totalWeight": 1921672664,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 56,
      "line": undefined,
      "name": "GHC.Conc.Signal.CAF",
      "selfWeight": 640,
      "totalWeight": 640,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 73,
      "line": undefined,
      "name": "GHC.IO.Encoding.CAF",
      "selfWeight": 2768,
      "totalWeight": 2768,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 75,
      "line": undefined,
      "name": "GHC.IO.Encoding.Iconv.CAF",
      "selfWeight": 200,
      "totalWeight": 200,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 84,
      "line": undefined,
      "name": "GHC.IO.Handle.FD.CAF",
      "selfWeight": 34704,
      "totalWeight": 34704,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 119,
      "line": undefined,
      "name": "Text.Printf.CAF",
      "selfWeight": 528,
      "totalWeight": 528,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 26,
      "line": undefined,
      "name": "Main.CAF:eta1_r6nI",
      "selfWeight": 0,
      "totalWeight": 2097152,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(21,1)-(38,42)",
      "key": 19,
      "line": undefined,
      "name": "Main.main",
      "selfWeight": 32,
      "totalWeight": 1921591544,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(56,1)-(57,42)",
      "key": 4,
      "line": undefined,
      "name": "Main.check",
      "selfWeight": 406149056,
      "totalWeight": 406149056,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:31:9-29",
      "key": 16,
      "line": undefined,
      "name": "Main.main.long",
      "selfWeight": 32,
      "totalWeight": 7864112,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 28,
      "line": undefined,
      "name": "Main.CAF:eta_r6nG",
      "selfWeight": 1096,
      "totalWeight": 1096,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 35,
      "line": undefined,
      "name": "Main.CAF:io1",
      "selfWeight": 1888,
      "totalWeight": 1888,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 27,
      "line": undefined,
      "name": "Main.CAF:lvl2_r6nH",
      "selfWeight": 0,
      "totalWeight": 7864112,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(61,1)-(63,26)",
      "key": 3,
      "line": undefined,
      "name": "Main.make",
      "selfWeight": 1512575520,
      "totalWeight": 1512575520,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:63:19-26",
      "key": 1,
      "line": undefined,
      "name": "Main.make.d2",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:63:9-16",
      "key": 2,
      "line": undefined,
      "name": "Main.make.i2",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 24,
      "line": undefined,
      "name": "Main.CAF:lvl4_r6nL",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 22,
      "line": undefined,
      "name": "Main.CAF:lvl8_r6nP",
      "selfWeight": 520,
      "totalWeight": 520,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 21,
      "line": undefined,
      "name": "Main.CAF:main1",
      "selfWeight": 16,
      "totalWeight": 16,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 29,
      "line": undefined,
      "name": "Main.CAF:main11",
      "selfWeight": 0,
      "totalWeight": 4194304,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:27:9-35",
      "key": 15,
      "line": undefined,
      "name": "Main.main.c",
      "selfWeight": 64,
      "totalWeight": 19922736,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 30,
      "line": undefined,
      "name": "Main.CAF:main12",
      "selfWeight": 0,
      "totalWeight": 15728432,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 23,
      "line": undefined,
      "name": "Main.CAF:main5",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 31,
      "line": undefined,
      "name": "Main.CAF:main7",
      "selfWeight": 880,
      "totalWeight": 880,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 25,
      "line": undefined,
      "name": "Main.CAF:main9",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:23:9-12",
      "key": 33,
      "line": undefined,
      "name": "Main.CAF:main_maxN",
      "selfWeight": 0,
      "totalWeight": 32,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:23:9-35",
      "key": 13,
      "line": undefined,
      "name": "Main.main.maxN",
      "selfWeight": 32,
      "totalWeight": 32,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:22:9",
      "key": 34,
      "line": undefined,
      "name": "Main.CAF:main_n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:22:9-14",
      "key": 12,
      "line": undefined,
      "name": "Main.main.n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:24:9-16",
      "key": 32,
      "line": undefined,
      "name": "Main.CAF:main_stretchN",
      "selfWeight": 0,
      "totalWeight": 32,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:24:9-27",
      "key": 14,
      "line": undefined,
      "name": "Main.main.stretchN",
      "selfWeight": 32,
      "totalWeight": 32,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:17:1-4",
      "key": 36,
      "line": undefined,
      "name": "Main.CAF:minN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:17:1-8",
      "key": 9,
      "line": undefined,
      "name": "Main.minN",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 146,
      "line": undefined,
      "name": "GC.GC",
      "selfWeight": 0,
      "totalWeight": 0,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 147,
      "line": undefined,
      "name": "PROFILING.OVERHEAD_of",
      "selfWeight": 2496,
      "totalWeight": 2496,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": 145,
      "line": undefined,
      "name": "SYSTEM.SYSTEM",
      "selfWeight": 34736,
      "totalWeight": 34736,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:35:26-54",
      "key": 18,
      "line": undefined,
      "name": "Main.main.\\\\",
      "selfWeight": 2224,
      "totalWeight": 46672,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:19:1-56",
      "key": 8,
      "line": undefined,
      "name": "Main.io",
      "selfWeight": 67912,
      "totalWeight": 67912,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:34:9-28",
      "key": 17,
      "line": undefined,
      "name": "Main.main.vs",
      "selfWeight": 0,
      "totalWeight": 1891637344,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(42,1)-(45,38)",
      "key": 11,
      "line": undefined,
      "name": "Main.depth",
      "selfWeight": 1120,
      "totalWeight": 1891637344,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:(49,1)-(52,31)",
      "key": 7,
      "line": undefined,
      "name": "Main.sumT",
      "selfWeight": 0,
      "totalWeight": 1891636224,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:51:9-31",
      "key": 6,
      "line": undefined,
      "name": "Main.sumT.a",
      "selfWeight": 1397760,
      "totalWeight": 945818112,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:52:9-31",
      "key": 5,
      "line": undefined,
      "name": "Main.sumT.b",
      "selfWeight": 1397760,
      "totalWeight": 945818112,
    },
    Frame {
      "col": undefined,
      "file": "src/Main.hs:45:9-38",
      "key": 10,
      "line": undefined,
      "name": "Main.depth.n",
      "selfWeight": 0,
      "totalWeight": 0,
    },
  ],
  "name": "binary-trees allocation",
  "stacks": Array [
    "MAIN.MAIN;GHC.Conc.Signal.CAF 640 B",
    "MAIN.MAIN;GHC.IO.Encoding.CAF 2.70 KB",
    "MAIN.MAIN;GHC.IO.Encoding.Iconv.CAF 200 B",
    "MAIN.MAIN;GHC.IO.Handle.FD.CAF 33.89 KB",
    "MAIN.MAIN;Text.Printf.CAF 528 B",
    "MAIN.MAIN;Main.CAF:eta1_r6nI;Main.main;Main.check 2.00 MB",
    "MAIN.MAIN;Main.CAF:eta1_r6nI;Main.main 32 B",
    "MAIN.MAIN;Main.CAF:eta_r6nG 1.07 KB",
    "MAIN.MAIN;Main.CAF:io1 1.84 KB",
    "MAIN.MAIN;Main.CAF:lvl2_r6nH;Main.main;Main.main.long;Main.make 7.50 MB",
    "MAIN.MAIN;Main.CAF:lvl2_r6nH;Main.main;Main.main.long 32 B",
    "MAIN.MAIN;Main.CAF:lvl8_r6nP 520 B",
    "MAIN.MAIN;Main.CAF:main1 16 B",
    "MAIN.MAIN;Main.CAF:main11;Main.main;Main.main.c;Main.check 4.00 MB",
    "MAIN.MAIN;Main.CAF:main11;Main.main;Main.main.c 32 B",
    "MAIN.MAIN;Main.CAF:main12;Main.main;Main.main.c;Main.make 15.00 MB",
    "MAIN.MAIN;Main.CAF:main12;Main.main;Main.main.c 32 B",
    "MAIN.MAIN;Main.CAF:main7 880 B",
    "MAIN.MAIN;Main.CAF:main_maxN;Main.main;Main.main.maxN 32 B",
    "MAIN.MAIN;Main.CAF:main_stretchN;Main.main;Main.main.stretchN 32 B",
    "MAIN.MAIN;PROFILING.OVERHEAD_of 2.44 KB",
    "MAIN.MAIN;SYSTEM.SYSTEM 33.92 KB",
    "MAIN.MAIN;Main.main;Main.main.\\\\;Main.io 43.41 KB",
    "MAIN.MAIN;Main.main;Main.main.\\\\ 2.17 KB",
    "MAIN.MAIN;Main.main;Main.io 22.91 KB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.check 190.67 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a;Main.make 710.00 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.a 1.33 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b;Main.check 190.67 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b;Main.make 710.00 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth;Main.sumT;Main.sumT.b 1.33 MB",
    "MAIN.MAIN;Main.main;Main.main.vs;Main.depth 1.09 KB",
    "MAIN.MAIN 648 B",
  ],
}
`;

exports[`importFromHaskell: indexToView 1`] = `0`;

exports[`importFromHaskell: profileGroup.name 1`] = `"binary-trees"`;
